### PR TITLE
WT-12440 Improve checkpoint prepare stats and progress logs

### PR DESCRIPTION
**Pull request by kommiharibabu**
Add more statistics in the code flow where the checkpoint prepare may be spending longer time than expected randomly and also add a progress log message to track the checkpoint prepare stage.

### CODE REVIEW

**Comment by kommiharibabu** [comment]
@quchenhao and @etienneptl please let me know if you think there are any other stats/logs that are required to identify the reason behind the checkpoint prepare operation delay.

File: dist/stat_data.py [hunk 1]
```diff
@@ -397,6 +397,10 @@ def __init__(self, name, desc, flags=''):
     CheckpointStat('checkpoint_handle_duration', 'most recent duration for gathering all handles (usecs)', 'no_clear,no_scale'),
     CheckpointStat('checkpoint_handle_duration_apply', 'most recent duration for gathering applied handles (usecs)', 'no_clear,no_scale'),
     CheckpointStat('checkpoint_handle_duration_skip', 'most recent duration for gathering skipped handles (usecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_handle_meta_check', 'most recent handles metadata checked'),
+    CheckpointStat('checkpoint_handle_meta_check_duration', 'most recent duration for handles metadata checked (usecs)', 'no_clear,no_scale'),
+    CheckpointStat('checkpoint_handle_lock', 'most recent handles metadata locked'),
+    CheckpointStat('checkpoint_handle_lock_duration', 'most recent duration for locking the handles (usecs)', 'no_clear,no_scale'),
```
Comment (quchenhao) [inline dist/stat_data.py:403]: I think the time most likely is spent in `__wt_checkpoint_get_handles`. Can we add some stats there?

File: src/txn/txn_ckpt.c [hunk 2]
```diff
@@ -1178,6 +1219,10 @@ __txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
     WT_WITH_SCHEMA_LOCK(session, ret = __checkpoint_prepare(session, &tracking, cfg));
     WT_ERR(ret);
 
+    /* Log the final checkpoint prepare progress message if needed. */
+    if (conn->ckpt_progress_msg_count > 0)
```
Comment (etienneptl) [inline src/txn/txn_ckpt.c:1223]: If the checkpoint is quick enough, this means we won't print this as `ckpt_progress_msg_count` should be `0`. Is that expected?

Comment (kommiharibabu) [reply src/txn/txn_ckpt.c:1223]: Yes, we do not want to log it for fast checkpoints.

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/txn/txn_ckpt.c [hunk 3]
```diff
@@ -478,7 +488,12 @@ __wt_checkpoint_get_handles(WT_SESSION_IMPL *session, const char *cfg[])
      * Decide whether the tree needs to be included in the checkpoint and if so, acquire the
      * necessary locks.
      */
+    time_start = __wt_clock(session);
     WT_SAVE_DHANDLE(session, ret = __checkpoint_lock_dirty_tree(session, true, force, true, cfg));
+    time_stop = __wt_clock(session);
+    time_diff = WT_CLOCKDIFF_US(time_stop, time_start);
+    ++S2C(session)->ckpt_lock;
+    S2C(session)->ckpt_lock_time += time_diff;
     WT_RET(ret);
```
Comment (quchenhao) [inline src/txn/txn_ckpt.c:497]: What about the call to `__wt_session_get_dhandle`? I reviewed the code and I think we cannot be blocked there but I may be wrong.

Comment (quchenhao) [reply src/txn/txn_ckpt.c:497]: If we want to save our time to debug this, can we also add stats inside `__checkpoint_lock_dirty_tree`?

Comment (kommiharibabu) [reply src/txn/txn_ckpt.c:497]: regarding timing stats for `__wt_session_get_dhandle ` inside the function `__wt_checkpoint_get_handles`, we are already gathering the stats for metadata check and lock tree. The remaining is the get dhandle. This timing stats should be recorded as part of the `checkpoint applied time`, so there is no need for specific stats to this.

Comment (kommiharibabu) [reply src/txn/txn_ckpt.c:497]: Added stats for `drop` also that is inside the `checkpoint_lock` function.

File: src/txn/txn_ckpt.c [hunk 4]
```diff
@@ -2014,6 +2061,11 @@ __checkpoint_lock_dirty_tree(
      */
     WT_ERR(__drop(session, NULL, ckptbase, name, strlen(name)));
```
Comment (etienneptl) [inline src/txn/txn_ckpt.c:2062]: Above this code, there is a loop where we try to drop specific checkpoints, check the following comment:
```
We may be dropping specific checkpoints, check the configuration.
```
We may want to increment the stat there too.

File: src/conn/conn_dhandle.c [hunk 5]
```diff
@@ -726,8 +726,15 @@ __wt_conn_btree_apply(WT_SESSION_IMPL *session, const char *uri,
             WT_STAT_CONN_SET(session, checkpoint_handle_skipped, conn->ckpt_skip);
```
Comment (etienneptl) [inline src/conn/conn_dhandle.c:726]: Can we A->Z this list based on the stat name variable?

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/conn/conn_dhandle.c (hunk 5 above)
Comment (kommiharibabu) [reply src/conn/conn_dhandle.c]: Fixed.

Comment (kommiharibabu) [reply src/conn/conn_dhandle.c]: Fixed in [3f03206](https://github.com/wiredtiger/wiredtiger/pull/10220/commits/3f032065de9a48e907f75fb11ba8f6ca4ef7c472)

**Review by etienneptl — APPROVED** [review]

**Pull request merged**